### DISCUSSION TITLE
post/handling-multi-channel-audio-in-naudio

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#67] [Announcements]
# post/handling-multi-channel-audio-in-naudio

Mark Heath's Development Blog

https://markheath.net/post/handling-multi-channel-audio-in-naudio

**Comment by timotech** [comment]
Hi Mark,
Thanks for the cool library. I want to use the MultiplexingWaveProvider for my case, but i'm a little confused. I have an 8 channel sound card. 4 left and 4 right.
I want to connect different radio stations to each making 8 radio stations and get recordings from the stations. My confusion is on mapping the channels, do I need to do a 1 to 1 mapping?
Such as 
waveProvider.ConnectInputToOutput(1,0);
waveProvider.ConnectInputToOutput(2,1);
and so on until I get to the 8th channel?

**Reply by markheath** [reply]
yes, although it's zero-based so I'd connect input 0 to output 0, input 1 to output 1 etc.

**Comment by timotech** [comment]
Hi Mark,Thanks for reply. I've been able to resolve the issue.
ThanksWarmest Regards
Timothy Ayodele
    On Monday, December 18, 2023 at 02:38:57 PM GMT+1, Mark Heath ***@***.***> wrote:  
 
 


yes, although it's zero-based so I'd connect input 0 to output 0, input 1 to output 1 etc.

**Comment by hoihoi000a** [comment]
Hi Mark,
I try to play the multiple channels in one audio.
WaveOutEvent(), WasapiOut(), DirectSoundOut() don't play the way I expect.
MultiplexingWaveProvider() this provider works well for me.
I could see that the output file matches my expectations.

**Comment by pgokcen** [comment]
Dear Mark;

I work on a project for multi zones and multi players using Dante Virtual Soundcard with .net core wpf project however I’m currently both developing the project and trying to learn at the same time, so it’s been challenging. The problem I encountered was that, when running Dante Virtual Soundcard in ASIO mode, the code I wrote to play audio from multiple players returned the error “Dante Virtual Soundcard already in use.” I researched this extensively and learned that an ASIO driver can only be initialized once per application. However, I have many audio players tied to Dante’s multiple channels, and they need to run concurrently yet independently. In my current implementation, I hear the same sound on every channel—they all start playing at once.

I’ve been trying hard to find a solution and came across your article. It seems I need to build a mixer architecture: convert MP3 files to PCM, write them into the mixer buffer, have the mixer monitor active players and pull in their data, and then route that data to the correct channels. Is it possible to implement this with NAudio? Could you help me? I’ve also looked into the ManagedBass library, but due to the lack of information I’m not ready to decide yet. Thanks in advance.

**Reply by markheath** [reply]
Yes it is an unfortunate limitation of ASIO. NAudio does have some capability with the "multiplexing" providers as I describe in the blog, but it certainly is not a straightforward undertaking. NAudio is probably overdue some improvements in this area, but sadly I don't have the time to work on it at the moment

**Reply by pgokcen** [reply]
Thank you for your reply Mark I am trying to overcome it with Bass enabling for each channel callback I hope it works ... I do not have any opinion right now coding still continue

**Comment by timotech** [comment]
Hi pgokcen,I was able to record multiple channels into different files using the normal waveInEvent. The only time I tried using a multiplexer was when I wanted to channel each waveInEvent to a speaker using BufferedWaveProvider like below.
//use buffered wave provider for getting speaker sound separate                BufferedWaveProvider waveIn = new(waveSource.WaveFormat);                speakers.Add(waveIn);But the trick is to declare a single recording function, and call it inside a loop that is handled by a task like below:for (int i = 0; i < stationCount; i++)                    {
_ = Task.Run(() =>                            {                                if (recordingFormat == ".wav")                                    RecordMicNAudio(deviceNum, stationName, prgBar, waveFileWriter, lblStatsRec, deviceChannel, staName);                                else                                    RecordMicNAudioMp3(deviceNum, stationName, prgBar, mp3FileWriter, lblStatsRec, deviceChannel, staName);                            });}
The code is not well formatted.However, there was a situation where I made a mistake, which I can't remember now, and all my left and right channel were recording into 1 sound. It may be difficult to put more into one sound but try using a loop and try if you can make use of my code below for extracting the audio and see if you can manipulate the extracted audio.
private byte[] ProcessAudioData(byte[] audioData, WaveFormat waveFormat, int option)        {            int bytesPerSample = waveFormat.BitsPerSample / 8;            int bytesPerFrame = waveFormat.Channels * bytesPerSample;
            if (option >= 2)            {                return audioData; // Record both channels            }            else if (option == 0)            {                return ExtractChannel(audioData, bytesPerSample, bytesPerFrame, 0); // Record left channel            }            else if (option == 1)            {                return ExtractChannel(audioData, bytesPerSample, bytesPerFrame, 1); // Record right channel            }            else            {                return new byte[0]; // Don't record either channel            }        }
        private byte[] ExtractChannel(byte[] stereoData, int bytesPerSample, int bytesPerFrame, int channelIndex)        {            int channelOffset = channelIndex * bytesPerSample;            byte[] extractedData = new byte[stereoData.Length / 2];
            for (int i = 0, j = 0; i < stereoData.Length; i += bytesPerFrame)            {                for (int b = 0; b < bytesPerSample; b++)                {                    extractedData[j++] = stereoData[i + channelOffset + b];                }            }
            return extractedData;        }

    On Monday, May 26, 2025 at 07:42:38 AM GMT+1, pgokcen ***@***.***> wrote:  
 
 


Thank you for your reply Mark I am trying to overcome it with Bass enabling for each channel callback I hope it works ... I do not have any opinion right now coding still continue

**Reply by pgokcen** [reply]
Hi timotech thank you for your reply and interest to help me on my issue the first problem I face with my code is my stream returns numbers like -2938466575858 I mean minus numbers is that normal ? And the output of my code is like Bass.Init → True, LastError: OK
[InitAsio] driver=1, rate=44100, channels=8, bytesPerSample=3
Bass.CreateStream returned: -2147483647, LastError: OK
[StartMixer] ASIO driver started. _isMixerRunning=True
#1: Ch=0
[OnAsioBuffer] Ch:0, Len:1536, T:10:17:14.158
  - mixing 1 sessions
#2: Ch=1
[OnAsioBuffer] Ch:1, Len:1536, T:10:17:14.178
  - mixing 1 sessions
#3: Ch=2
[OnAsioBuffer] Ch:2, Len:1536, T:10:17:14.186
  - mixing 1 sessions
    > playerId=1 writes to Ch=2
[Decoder] ReadNextChunk -> 0 bytes
The thread '.NET TP Worker' (16084) has exited with code 0 (0x0).
The thread '.NET TP Worker' (6848) has exited with code 0 (0x0).
The thread '.NET TP Worker' (21696) has exited with code 0 (0x0).
#4: Ch=3
[OnAsioBuffer] Ch:3, Len:1536, T:10:17:28.541
  - mixing 1 sessions
    > playerId=1 writes to Ch=3
[Decoder] ReadNextChunk -> 0 bytes
#5: Ch=4
[OnAsioBuffer] Ch:4, Len:1536, T:10:17:28.552
  - mixing 1 sessions
#6: Ch=5
[OnAsioBuffer] Ch:5, Len:1536, T:10:17:28.559
  - mixing 1 sessions
#7: Ch=6
[OnAsioBuffer] Ch:6, Len:1536, T:10:17:28.566
  - mixing 1 sessions
#8: Ch=7
[OnAsioBuffer] Ch:7, Len:1536, T:10:17:28.573
  - mixing 1 sessions
#9: Ch=0
[OnAsioBuffer] Ch:0, Len:1536, T:10:17:28.592
  - mixing 1 sessions
#10: Ch=1
[OnAsioBuffer] Ch:1, Len:1536, T:10:17:28.600
  - mixing 1 sessions
#11: Ch=2
[OnAsioBuffer] Ch:2, Len:1536, T:10:17:28.607
  - mixing 1 sessions
    > playerId=1 writes to Ch=2
[Decoder] ReadNextChunk -> 0 bytes
#12: Ch=3 ........ is going like that and no sound it is very difficult to post all my code here I am stuck I must get sound :( 

var asioDriver = new BassAsioDriver();
var mixer = new Mixer();
var engine = new AudioEngine(asioDriver, mixer);


int driverIndex = 1;


await engine.InitializeAsio(driverIndex, 44100, 8);


var session = new DecoderPlaybackSession(@"C:\music\test.mp3");

// 5) Stereo için iki ASIO kanalı (TX3 ve TX4 → index 2 ve 3)
engine.AddSession(
    playerId: 1,
    session: session,
    outputChannelIndices: new[] { 2, 3 }
);


await engine.StartMixerAsync();


await Task.Delay(TimeSpan.FromSeconds(30));

// 8) Durdur ve temizle
await engine.StopMixerAsync();
session.Dispose();
await engine.ShutdownAsync();
